Add support for web extension registered content scripts

kiaraarose opened the pull request:
#### 309fb63df40fb69a38444e281ab03d02073832c0
<pre>
Add support for web extension registered content scripts
<a href="https://bugs.webkit.org/show_bug.cgi?id=261769">https://bugs.webkit.org/show_bug.cgi?id=261769</a>

Reviewed by Timothy Hatcher.

This patch adds support for registered web extension content scripts with the scriptingAPI. To do this,
expand the WebExtensionDynamicScripts namespace to include a new WebExtensionRegisteredScript class
that will keep track of the scripts added with the API. This class will also keep track of the
user scripts and stylesheets added, and removes them when a script is either updated or unregistered.

To do: Allow the scripts to persist across browser sessions if &apos;persistAcrossSessions&apos; is true.
In order to do this, we first need to add support for web extension storage, which is being tracked
in <a href="https://bugs.webkit.org/show_bug.cgi?id=264892.">https://bugs.webkit.org/show_bug.cgi?id=264892.</a>

* Source/WebKit/Shared/Extensions/WebExtensionDynamicScripts.serialization.in:
* Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptParameters.h:
(): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
(WebKit::WebExtensionContext::scriptingRegisterScripts):
(WebKit::WebExtensionContext::scriptingUpdateRegisteredScripts):
(WebKit::WebExtensionContext::scriptingGetRegisteredScripts):
(WebKit::WebExtensionContext::scriptingUnregisterContentScripts):
(WebKit::WebExtensionContext::parseRegisteredContentScripts):
Validate scripts and create InjectedContentData for each one.

(WebKit::WebExtensionContext::scriptingUnregisterScripts): Deleted.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::populateContentScriptPropertiesIfNeeded):

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::addInjectedContent):
Keep track of user scripts and style sheets for registered scripts.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
(WebKit::WebExtensionDynamicScripts::WebExtensionRegisteredScript::updateParameters):
(WebKit::WebExtensionDynamicScripts::WebExtensionRegisteredScript::merge):
Take the existing properties if they&apos;re not specified in the update parameters.

(WebKit::WebExtensionDynamicScripts::WebExtensionRegisteredScript::addUserScript):
(WebKit::WebExtensionDynamicScripts::WebExtensionRegisteredScript::addUserStyleSheet):
Keep track of user scripts and style sheets for registered scripts.

(WebKit::WebExtensionDynamicScripts::WebExtensionRegisteredScript::removeUserScriptsAndStyleSheets):
(WebKit::WebExtensionDynamicScripts::WebExtensionRegisteredScript::removeUserScripts):
(WebKit::WebExtensionDynamicScripts::WebExtensionRegisteredScript::removeUserStyleSheets):
Iterate through all user content controllers and remove all the user scripts and stylesheets for the
given script ID.

* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionDynamicScripts.h:
(WebKit::WebExtensionDynamicScripts::WebExtensionRegisteredScript::create):
(WebKit::WebExtensionDynamicScripts::WebExtensionRegisteredScript::WebExtensionRegisteredScript):
(WebKit::WebExtensionDynamicScripts::WebExtensionRegisteredScript::parameters const):

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
(WebKit::toWebAPI):
(WebKit::WebExtensionAPIScripting::registerContentScripts):
(WebKit::WebExtensionAPIScripting::getRegisteredContentScripts):
(WebKit::WebExtensionAPIScripting::updateContentScripts):
(WebKit::WebExtensionAPIScripting::unregisterContentScripts):

(WebKit::WebExtensionAPIScripting::validateRegisteredScripts):
Fix bug caused by how we check for the the injectionTime.

(WebKit::WebExtensionAPIScripting::parseRegisteredContentScripts):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIScripting.h:

Canonical link: <a href="https://commits.webkit.org/271318@main">https://commits.webkit.org/271318@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/86dc3fcd53e364b52d20b03674e6fc0156b2d997

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27994 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6632 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29298 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/30524 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25530 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/28489 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8630 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4022 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/25293 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28261 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5383 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24046 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4638 "Found 1 new API test failure: /WPE/TestWebKitWebView:/webkit/WebKitWebView/notification (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4816 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25054 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/31213 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25594 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25489 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31078 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4823 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2989 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28889 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6360 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/5267 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3621 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5313 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/UIProcess/Extensions/WebExtensionDynamicScripts.h:
```diff
@@ -66,6 +75,46 @@ class InjectionResultHolder : public RefCounted<InjectionResultHolder> {
+        return adoptRef(*new WebExtensionRegisteredScript(std::forward<Args>(args)...));
+    }
+
+    explicit WebExtensionRegisteredScript(WebExtensionContext& context, String identifier, std::optional<Vector<String>> css, std::optional<Vector<String>> js, std::optional<InjectionTime> injectionTime, std::optional<Vector<String>> excludeMatchPatterns, std::optional<Vector<String>> matchPatterns, std::optional<bool> allFrames, std::optional<bool> persistent, std::optional<bool> forMainWorld)
```
had initially tried to pass in the parameters directly but ran into issues. i see that @xeenon did it successfully in his latest PR so will try that approach again

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -142,32 +142,196 @@
+void WebExtensionContext::scriptingRegisterContentScripts(const Vector<WebExtensionRegisteredScriptParameters>& scripts, CompletionHandler<void(WebExtensionDynamicScripts::Error)>&& completionHandler)
 {
-    // FIXME: <https://webkit.org/b/261769> Implement this.
+    NSString *apiName = @"scripting.registerContentScripts()";
```
`static NSString * const apiName`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -142,32 +142,196 @@
 void WebExtensionContext::scriptingUpdateRegisteredScripts(const Vector<WebExtensionRegisteredScriptParameters>& scripts, CompletionHandler<void(WebExtensionDynamicScripts::Error)>&& completionHandler)
 {
-    // FIXME: <https://webkit.org/b/261769> Implement this.
+    NSString *apiName = @"scripting.updateContentScripts()";
```
`static NSString * const apiName`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -142,32 +142,196 @@
+    NSString *apiName = @"scripting.updateContentScripts()";
+    Vector<WebExtensionRegisteredScriptParameters> updatedParameters;
+
+    for (auto parameters : scripts) {
```
`auto&` to avoid a copy.

kiaraarose replied:
i didn't add `auto&` here since the parameters are being modified in the `merge` method

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -142,32 +142,196 @@
+
+    for (auto& parameters : updatedParameters) {
+        auto scriptID = parameters.identifier;
+        RefPtr registeredScript = m_registeredScriptsMap.get(scriptID);
```
Does this need a null check / continue or ASSERT?

kiaraarose replied:
at this point the registered map should contain the script ID, which we verify on line 174 but can't hurt to add an assert

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -142,32 +142,196 @@
+        for (auto& entry : m_registeredScriptsMap)
+            results.append(entry.value.get().parameters());
+    } else {
+        for (auto scriptID : scriptIDs) {
```
`auto&`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -142,32 +142,196 @@
-    completionHandler({ }, std::nullopt);
+    if (scriptIDs.isEmpty()) {
+        // Removes all registered scripts if no filter is specififed.
+        for (auto scriptID : copyToVector(m_registeredScriptsMap.keys()))
```
`auto&`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -142,32 +142,196 @@
+        return;
+    }
+
+    for (auto scriptID : scriptIDs) {
```
`auto&`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -142,32 +142,196 @@
+        }
+    }
+
+    for (auto scriptID : scriptIDs)
```
`auto&`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -142,32 +142,196 @@
+        });
+
+        for (NSString *scriptPath in scriptPaths) {
+            if (!extension().resourceStringForPath(scriptPath)) {
```
Might want to pass `SuppressNotFoundErrors::Yes` to avoid logging the error since we report it another way here.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -142,32 +142,196 @@
+        auto *matchesArray = parameters.matchPatterns ? createNSArray(parameters.matchPatterns.value()).get() : @[ ];
+        for (NSString *matchPatternString in matchesArray) {
+            if (!matchPatternString.length) {
+                *errorMessage = (toErrorString(callingAPIName, nil, @"script with ID '%@' contains an empty match pattern", (NSString *)scriptID));
```
Drop extra parens.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -142,32 +142,196 @@
+
+            auto matchPattern = WebExtensionMatchPattern::getOrCreate(matchPatternString);
+            if (!matchPattern || !matchPattern->isSupported()) {
+                *errorMessage = (toErrorString(callingAPIName, nil, @"script with ID '%@' has an invalid match pattern '%@'", (NSString *)scriptID, matchPatternString));
```
Drop extra parens.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -142,32 +142,196 @@
+        auto *excludeMatchesArray = parameters.excludeMatchPatterns ? createNSArray(parameters.excludeMatchPatterns.value()).get() : @[ ];
+        for (NSString *matchPatternString in excludeMatchesArray) {
+            if (!matchPatternString.length) {
+                *errorMessage = (toErrorString(callingAPIName, nil, @"script with ID '%@' contains an empty exclude match pattern", (NSString *)scriptID));
```
Drop extra parens.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -142,32 +142,196 @@
+
+            auto matchPattern = WebExtensionMatchPattern::getOrCreate(matchPatternString);
+            if (!matchPattern || !matchPattern->isSupported()) {
+                *errorMessage = (toErrorString(callingAPIName, nil, @"script with ID '%@' has an invalid exclude match pattern '%@'", (NSString *)scriptID, matchPatternString));
```
Drop extra parens.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -142,32 +142,196 @@
+                return false;
+            }
+
+            auto matchPattern = WebExtensionMatchPattern::getOrCreate(matchPatternString);
```
`RefPtr` instead of `auto`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -142,32 +142,196 @@
+                return false;
+            }
+
+            auto matchPattern = WebExtensionMatchPattern::getOrCreate(matchPatternString);
```
`RefPtr` instead of `auto`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -1850,7 +1850,7 @@ static bool parseCommandShortcut(const String& shortcut, OptionSet<ModifierFlags
             recordError(createError(Error::InvalidContentScripts, WEB_UI_STRING("Manifest `content_scripts` entry has unknown `run_at` value.", "WKWebExtensionErrorInvalidContentScripts description for unknown 'run_at' value")));
 
-        m_staticInjectedContents.append({ WTFMove(includeMatchPatterns), WTFMove(excludeMatchPatterns), injectionTime, matchesAboutBlank, injectsIntoAllFrames, false, scriptPaths, styleSheetPaths, includeGlobPatternStrings, excludeGlobPatternStrings });
+        m_staticInjectedContents.append({ WTFMove(includeMatchPatterns), WTFMove(excludeMatchPatterns), injectionTime, ""_s, matchesAboutBlank, injectsIntoAllFrames, false, scriptPaths, styleSheetPaths, includeGlobPatternStrings, excludeGlobPatternStrings });
```
As this gets longer, this might be better as a var that you WTFMove here, so you can set the struct properties by name.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -196,6 +193,89 @@ WebExtensionScriptInjectionResultParameters toInjectionResultParameters(id resul
+    if (m_userScriptsMap.contains(identifier))
+        m_userScriptsMap.get(identifier).append(userScript);
+    else
+        m_userScriptsMap.add(identifier, UserScriptVector { userScript });
```
This can be done with `ensure()`.

```
auto& userScripts = m_userScriptsMap.ensure(identifier, [&] {
    return UserScriptVector { };
}).iterator->value;
userScripts.append(userScript);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -196,6 +193,89 @@ WebExtensionScriptInjectionResultParameters toInjectionResultParameters(id resul
+    if (m_userStyleSheetsMap.contains(identifier))
+        m_userStyleSheetsMap.get(identifier).append(userStyleSheet);
+    else
+        m_userStyleSheetsMap.add(identifier, UserStyleSheetVector { userStyleSheet });
```
Ditto with `ensure()`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -196,6 +193,89 @@ WebExtensionScriptInjectionResultParameters toInjectionResultParameters(id resul
+void WebExtensionRegisteredScript::removeUserScripts(String identifier)
+{
+    auto userScripts = m_userScriptsMap.take(identifier);
+    auto allUserContentControllers = m_extensionContext.get()->extensionController()->allUserContentControllers();
```
Shouldn't need `.get()`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -196,6 +193,89 @@ WebExtensionScriptInjectionResultParameters toInjectionResultParameters(id resul
+void WebExtensionRegisteredScript::removeUserStyleSheets(String identifier)
+{
+    auto userStyleSheets = m_userStyleSheetsMap.take(identifier);
+    auto allUserContentControllers = m_extensionContext.get()->extensionController()->allUserContentControllers();
```
Shouldn't need `.get()`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -196,6 +193,89 @@ WebExtensionScriptInjectionResultParameters toInjectionResultParameters(id resul
+    m_parameters = parameters;
+}
+
+void WebExtensionRegisteredScript::merge(WebExtensionRegisteredScriptParameters& parameters)
```
`const WebExtensionRegisteredScriptParameters&` since you don't modify it.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -196,6 +193,89 @@ WebExtensionScriptInjectionResultParameters toInjectionResultParameters(id resul
+
+}
+
+void WebExtensionRegisteredScript::updateParameters(WebExtensionRegisteredScriptParameters& parameters)
```
`const WebExtensionRegisteredScriptParameters&` since you don't modify it.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -196,6 +193,89 @@ WebExtensionScriptInjectionResultParameters toInjectionResultParameters(id resul
+        parameters.forMainWorld = m_parameters.forMainWorld.value();
+}
+
+void WebExtensionRegisteredScript::addUserScript(String identifier, API::UserScript& userScript)
```
`const String& identifier`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -196,6 +193,89 @@ WebExtensionScriptInjectionResultParameters toInjectionResultParameters(id resul
+        m_userScriptsMap.add(identifier, UserScriptVector { userScript });
+}
+
+void WebExtensionRegisteredScript::addUserStyleSheet(String identifier, API::UserStyleSheet& userStyleSheet)
```
`const String& identifier`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -196,6 +193,89 @@ WebExtensionScriptInjectionResultParameters toInjectionResultParameters(id resul
+        m_userStyleSheetsMap.add(identifier, UserStyleSheetVector { userStyleSheet });
+}
+
+void WebExtensionRegisteredScript::removeUserScriptsAndStyleSheets(String identifier)
```
`const String& identifier`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -196,6 +193,89 @@ WebExtensionScriptInjectionResultParameters toInjectionResultParameters(id resul
+    removeUserStyleSheets(identifier);
+}
+
+void WebExtensionRegisteredScript::removeUserScripts(String identifier)
```
`const String& identifier`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -196,6 +193,89 @@ WebExtensionScriptInjectionResultParameters toInjectionResultParameters(id resul
+    auto userScripts = m_userScriptsMap.take(identifier);
+    auto allUserContentControllers = m_extensionContext.get()->extensionController()->allUserContentControllers();
+
+    for (Ref userScript : userScripts) {
```
`auto&` is fine for loops.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -196,6 +193,89 @@ WebExtensionScriptInjectionResultParameters toInjectionResultParameters(id resul
+    auto userStyleSheets = m_userStyleSheetsMap.take(identifier);
+    auto allUserContentControllers = m_extensionContext.get()->extensionController()->allUserContentControllers();
+
+    for (Ref userStyleSheet : userStyleSheets) {
```
Ditto. `auto&`

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -119,6 +119,49 @@
+        return documentEnd;
+    if (injectionTime == WebExtension::InjectionTime::DocumentIdle)
+        return documentIdle;
+    return documentStart;
```
A switch here would help catch new cases that need added.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -211,11 +254,8 @@
-            callback->reportError(error.value());
-        else
-            callback->call();
+    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::ScriptingGetRegisteredScripts(WTFMove(scriptIDs)), [protectedThis = Ref { *this }, callback = WTFMove(callback)](Vector<WebExtensionRegisteredScriptParameters> scripts) {
```
`const Vector<WebExtensionRegisteredScriptParameters>&`

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -192,9 +235,9 @@
 
     Vector<WebExtensionRegisteredScriptParameters> parameters;
-    parseRegisteredContentScripts(scripts, parameters);
+    parseRegisteredContentScripts(scripts, true, parameters);
```
An `enum class : bool { No, Yes }` for this would help make the call site clear what true / false means.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -533,11 +570,19 @@
+        if (NSString *world = script[worldKey])
+            parameters.forMainWorld = [world isEqualToString:mainWorld];
+        else if (isRegisteringScript)
+            parameters.forMainWorld = false;
```
I think `WebExtensionContentWorldType` is still better than a `forMainWorld` `bool`.

kiaraarose replied:
yea I was going back and forth between what to use here. `WebExtensionContentWorldType` is probably better in the case other worlds are added

kiaraarose committed:
Add support for web extension registered content scripts
https://bugs.webkit.org/show_bug.cgi?id=261769

Reviewed by Timothy Hatcher.

This patch adds support for registered web extension content scripts with the scriptingAPI. To do this,
expand the WebExtensionDynamicScripts namespace to include a new WebExtensionRegisteredScript class
that will keep track of the scripts added with the API. This class will also keep track of the
user scripts and stylesheets added, and removes them when a script is either updated or unregistered.

To do: Allow the scripts to persist across browser sessions if 'persistAcrossSessions' is true.
In order to do this, we first need to add support for web extension storage, which is being tracked
in https://bugs.webkit.org/show_bug.cgi?id=264892.

* Source/WebKit/Shared/Extensions/WebExtensionDynamicScripts.serialization.in:
* Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptParameters.h:
(): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
(WebKit::WebExtensionContext::scriptingRegisterScripts):
(WebKit::WebExtensionContext::scriptingUpdateRegisteredScripts):
(WebKit::WebExtensionContext::scriptingGetRegisteredScripts):
(WebKit::WebExtensionContext::scriptingUnregisterContentScripts):
(WebKit::WebExtensionContext::parseRegisteredContentScripts):
Validate scripts and create InjectedContentData for each one.

(WebKit::WebExtensionContext::scriptingUnregisterScripts): Deleted.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::populateContentScriptPropertiesIfNeeded):

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::addInjectedContent):
Keep track of user scripts and style sheets for registered scripts.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
(WebKit::WebExtensionDynamicScripts::WebExtensionRegisteredScript::updateParameters):
(WebKit::WebExtensionDynamicScripts::WebExtensionRegisteredScript::merge):
Take the existing properties if they're not specified in the update parameters.

(WebKit::WebExtensionDynamicScripts::WebExtensionRegisteredScript::addUserScript):
(WebKit::WebExtensionDynamicScripts::WebExtensionRegisteredScript::addUserStyleSheet):
Keep track of user scripts and style sheets for registered scripts.

(WebKit::WebExtensionDynamicScripts::WebExtensionRegisteredScript::removeUserScriptsAndStyleSheets):
(WebKit::WebExtensionDynamicScripts::WebExtensionRegisteredScript::removeUserScripts):
(WebKit::WebExtensionDynamicScripts::WebExtensionRegisteredScript::removeUserStyleSheets):
Iterate through all user content controllers and remove all the user scripts and stylesheets for the
given script ID.

* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionDynamicScripts.h:
(WebKit::WebExtensionDynamicScripts::WebExtensionRegisteredScript::create):
(WebKit::WebExtensionDynamicScripts::WebExtensionRegisteredScript::WebExtensionRegisteredScript):
(WebKit::WebExtensionDynamicScripts::WebExtensionRegisteredScript::parameters const):

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
(WebKit::toWebAPI):
(WebKit::WebExtensionAPIScripting::registerContentScripts):
(WebKit::WebExtensionAPIScripting::getRegisteredContentScripts):
(WebKit::WebExtensionAPIScripting::updateContentScripts):
(WebKit::WebExtensionAPIScripting::unregisterContentScripts):

(WebKit::WebExtensionAPIScripting::validateRegisteredScripts):
Fix bug caused by how we check for the the injectionTime.

(WebKit::WebExtensionAPIScripting::parseRegisteredContentScripts):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIScripting.h:

Canonical link: https://commits.webkit.org/271318@main
